src/Disks/DiskSelector.cpp

```diff
@@ -77,6 +77,9 @@ DiskSelectorPtr DiskSelector::updateFromConfig(
 
             /// TODO: Ideally ClickHouse shall complain if disk has changed, but
             /// implementing that may appear as not trivial task.
+            auto disk = std::static_pointer_cast<DiskLocal>(result->getDisksMap().find(disk_name)->second);
+            if (disk)
+                disk->updateFromConfig(config, disk_config_prefix, context);
```

Comment:
Maybe this should be done in background?
Like other config directives reloading - https://github.com/ClickHouse/ClickHouse/blob/ccb7e6e03f75f243c956e2718f615fab344c0e64/programs/server/Server.cpp#L685

Or, maybe config should include some version, so that you can check against it, and do nothing if version is the same.

Also it should note that it will not work if DiskPtr is cached, and this includes lot's of engines (Distriubted, TInyLog, Log, ....), but AFAICS this will work for MergeTree engines.

Reply:
1. I think this reload is already running  in background. Invoking chain like below: 
```
main_config_reloader
-> Context::updateStorageConfiguration
-> DiskSelector::updateFromConfig
-> DiskLocal::updateFromConfig
```

2. Decide whether to reload by version is a good idea.  I will finish this optimization. 
3. It is `DiskPtr` that is referenced by engines. If the object which `DiskPtr` pointed to just updated from config, the updating is visible for engines.

Reply:
>I think this reload is already running in background. Invoking chain like below:

Yep, but I though about updating it just from there, but forget it, config version will be enough

> It is DiskPtr that is referenced by engines. If the object which DiskPtr pointed to just updated from config, the updating is visible for engines.

You are right (I though that your patch creates new Disk)

Also have you considered not only disks but also volumes?

Reply:
I have not thought about volumns yet. 

In the first step, I planed to enable disk reload(not only `DiskLocal`, but also `DiskMemory` `DiskS3` and so on).  
Then considered reloading of volumns.  
This looks like a huge plan, I will try to finish it in this weekend.

Reply:
>It is DiskPtr that is referenced by engines. If the object which DiskPtr pointed to just updated from config, the updating is visible for engines.

Also note that if such storage (Log or similar) uses some unique volume, for which none will call updateFromConfig it will not be updated, that's why I was suggesting to update them in background (so that you explicitly iterate over all existing disks to update them all periodically, and in this case you may not need config version, since this will add overhead only into this background update)

Reply:
> Also note that if such storage (Log or similar) uses some unique volume, for which none will call updateFromConfig it will not be updated, that's why I was suggesting to update them in background (so that you explicitly iterate over all existing disks to update them all periodically, and in this case you may not need config version, since this will add overhead only into this background update)

Sorry, besides no version compare,  I still didn't  understand the advantage of background schedule over another solution. Disk reload(at least for `DiskLocal`) is not a heavy operation, and I think putting it into `main_config_reloader` thread is not a big problem...

Reply:
Since I didn't fully understand the points of  background schedule  Maybe I should try version solution firstly...

Reply:
> I still didn't understand the advantage of background schedule over another solution.

Let's consider the following example:

- One external disk:

```xml
<storage_configuration>
  <disks>
    <external_for_log_data_table>
      <path>/tmp/</path>
      <keep_free_space_bytes>1000</keep_free_space_bytes>
    </external_for_log_data_table>
  </disks>
</storage_configuration>
```

- Log table, that uses this disk:

```sql
create table data (key Int) engine=Log() settings disk='external_for_log_data_table'
```

And this will create StorageLog object, with cached Disk, that will be never updated, until someone will get it via DiskSelector (or similar API).

But if you will update disks in background (i.e. call `DiskLocal::updateFromConfigIfChanged` for each existing disk somehow), then you don't have such problem, and the changes will be applied to such cached Disks too (for whom nobody call getDisk again).

Sure you can workaround the problem by detach/attach, or create another table with the same disk name to call update, but this does not looks sane

Reply:
I think StorageLog will get the updated disk after reloading disk config. See code: 
```
registerStorageLog -> Context::getDisk -> Context::getDiskSelector -> DiskSelector::get -> StorageLog::create  
```
`DiskSelector` contains all the disks defined in config files, including the disk used by `StorageLog`.  In this PR, when I execute `DiskSelector::updateFromConfig`, all the shard disk objects are updated if changed.

Reply:
@taiyang-li Yes, you are right (I though about different call chain by some reason).

BTW looks like it worth adding some test for this.

Reply:
Test already added.